Make auto-creating encoding tickets configurable + allow creating additional encoding tickets via API

pegro opened the pull request:
See #90 
- create_missing_encoding_tickets is never called with a profile_id, so remove this parameter
- allow RPC API to create additional encoding tickets, even multiple tickets of the same encoding profile (and version)
- Make auto-creation of encoding tickets configurable in project settings (default auto-creation is enabled)
- Allow setting priority and auto-creation setting when assigning new encoding profiles to a project

jjeising commented on src/Application/Controller/Projects.php:
```diff
@@ -80,14 +80,18 @@ public function profiles() {
+								'encoding_profile_version_id' => $values['add']['version'],
+								'priority' => $values['add']['priority'],
+								'auto_create' => $values['add']['auto_create']
+							]
```
When `$values['add']` is an array, it should work to change `version` in the template to `encoding_profile_version_id` and just use

```php
$this->project->save([
	'EncodingProfileVersion' => [
		$values['add']
	]
]);
```

jjeising commented on src/Application/Controller/Projects.php:
```diff
@@ -114,6 +118,25 @@ public function profiles() {
+				
+				if (!empty($values['auto_create']) and
+					is_array($values['auto_create'])) {
+					foreach ($values['auto_create'] as $id => $auto_create) {
```
We normally use camel case for variables (except in XMLRPC_Handler).

pegro replied:
Okay.

jjeising commented on src/Application/Controller/Projects.php:
```diff
@@ -114,6 +118,25 @@ public function profiles() {
+						if (empty($auto_create['1']) && !$auto_create['0'] or
+							$auto_create['1'] === $auto_create['0']) {
+							continue;
+						}
```
Is there any reason to update this only when it has changed?

pegro replied:
Save some unnecessary queries ;)
But I also wasn't able to disable all profiles, since when all checkboxes are unchecked, the $values['auto_create'] key doesn't exist and I didn't want to assume this means all profiles have auto_create = false. I fixed that using a hidden field, but then went ahead and added those conditions to the if-clause.

jjeising replied:
> Save some unnecessary queries ;)

We already have those for priority and this looks very manual, but it's probably fine.

> since when all checkboxes are unchecked, the $values['auto_create'] key doesn't exist and I didn't want to assume this means all profiles have auto_create = false

I don't see any case where this would cause issues?

pegro replied:
How would solve that? Than I need a disableAutoCreateOnAllProfilesForProject() function in the model or another way to get iterate over all profiles to update them all.
I could use $values['priority'], but this feels messy.

jjeising replied:
> How would you solve that?

If you move these all values to a common array (e.g. `profiles[id][auto_create]`) they should be always filled, because of [Form](https://github.com/jjeising/framework/blob/crs-tools-tracker/Library/Form.php#L158). This currently does not seem to first if the first field is already missing (because of the `index === $maxDepth` condition), maybe I can patch that.

jjeising commented on src/Application/Controller/XMLRPC/Handler.php:
```diff
@@ -376,22 +376,29 @@ public function createEncodingTicket($meta_ticket_id, $encoding_profile_id, arra
-				// get ticket
-				$ticket = Ticket::find(['id' => $ticket_id]);
+				// create encoding ticket
+				$ticket = Ticket::createOrThrow($ticket_data);
```
When `$ticket_data` is not needed anywhere else I would probably prefer using an inline array as argument.

pegro replied:
Sure.

pegro commented:
Uhm, you can delete tickets using the UI. Or do you mean deletion via API?

pegro commented:
Yeah, there is no bulk-remove action, which would solve that too.

But the idea of "Remove untouched" is good.

pegro commented:
Regarding that "remove untouched": removing only tickets of the given profile **version** or all untouched of this **profile**?

jjeising commented:
> Regarding that "remove untouched": removing only tickets of the given profile version or all untouched of this profile?

Good question. I think all of the selected encoding profile might be a good idea. Otherwise a user might need to change the encoding profile version for cleanup.

For _untouched_ `created = modified` should be enough I think?

pegro commented:
What about ```state = inital_state``` ?
Or maybe just a "show tickets" link pointing to the ticket list in bulk mode with the given profile pre-selected, and adding a bulk delete action? Since untouched isn't that useful. Maybe you decide after some hours of the event you want to use a completely different profile and remove all (even already processed) encoding tickets of a profile?! Is that a use-case to consider?

a-tze commented:
> Maybe you decide after some hours of the event you want to use a completely different profile and remove all (even already processed) encoding tickets of a profile?! Is that a use-case to consider?

Erm...  isnt this done by just removing that encoding profile?

pegro commented:
Mh right.
So this is just about the automatically created ones.

jjeising commented:
> So this is just about the automatically created ones.

My thought was: User clicks _Auto create_, form is autosubmitted, lots of tickets are created. User unticks the box and is unhappy.

a-tze commented on src/Application/Model/Ticket.php:
```diff
@@ -616,15 +616,6 @@ public static function createMissingEncodingTickets($project) {
 			return $handle->fetch()['create_missing_encoding_tickets'];
 		}
 		
-		public static function createMissingEncodingTicket($metaTicket, $encodingProfile) {
```
This method is used when adding an encoding profile to the project in the current behaviour (Controller/Projects.php, around line 91).
Am I missing where the auto create will take place if it is enabled?

pegro replied:
That's another reason why I want to remove that method. It's so easy to mistake this method as it's plural (Ticket**s**) version right above in the file.

a-tze replied:
Got me, I used only full text search and was confused though I did see both versions of the method.

pegro commented:
PR updated
- rebased on top of master
- fixed changing versions after commit ￼"Project: Allow setting priority and auto-create setting when assignin…"
- added auto-remove of untouched (created = modified) tickets of given profile when disabling auto-create

a-tze commented on src/Application/Model/Ticket.php:
```diff
@@ -617,16 +617,7 @@ public static function createMissingEncodingTickets($project) {
 			return $handle->fetch()['create_missing_encoding_tickets'];
 		}
 		
-		public static function createMissingEncodingTicket($metaTicket, $encodingProfile) {
```
Something went wrong on this line. The (untouched) function is not named `getDependingEncodingTicketState` anymore, but `getDependeeTicketState`

pegro replied:
Oh, right.

jjeising commented:
> Being able to have multiple tickets of the same encoding profile (version) has some obvious impact on view_servicable_tickets and other database objects that handle profile dependency.

Maybe I missed this – why would we want that?

a-tze commented:
It's explained a little bit in 70fe3a7 . 

This also got me red-handed. If asked in the middle of the night I would swear that there is a uniqueness constraint like (project_id, encoding_profile_version_id) but obviously there has never been such a restriction (at least I couldn find one in the history) which really surprised me.

pegro commented:
But I agree, without those parameters they would be the same.
The UI doesn't allow adding the same profile **version** twice, if I read the code correctly. So this is only relevant for tickets created via the API.

a-tze commented:
Ok, makes sense.
Any ideas where to put a "YMMV" warning for this situation? I just thought about how a constraint could look like, but it's not as easy as I initially thought.

pegro commented:
One idea: disallow having multiple tickets of a profile, which is used as a dependee. Does that help?

a-tze commented:
> The UI doesn't allow adding the same profile version twice

No, it does not even allow adding the same profile twice.

> One idea: disallow having multiple tickets of a profile, which is used as a dependee. Does that help?

Yes that was the idea, but I just think about if it's doable in the relation project <-> EPV, of if it must be done on the ticket level.

a-tze commented:
> Maybe we can add a special handling for <xsl:include href="[encodingprofileversionid].xls"/> or similar?

Sounds nice but I think this is not very handy. You split your things across multiple profiles, while just having a profile which is controlled by properties is so much simpler and more transparent.

a-tze commented:
> For now I would check it in the API Handler. 

I'm fine with that.

pegro commented:
PR updated.
- fixed broken function name due to rebasing
- added duplicate prevention for "master" encoding tickets

a-tze commented:
IMO this can be merged.

pegro commented:
Anyone want to hit the Approve button?
Or shall I just merge it

a-tze commented:
The thing is, I dont see an approve button (anymore?) I think it disappears if review is requested and comments or discussion is made. Just do it...

pegro commented:
Don't you have a "Review changes" button when you go to https://github.com/crs-tools/tracker/pull/172/files?
